File: pool/v/object_pool.java
import java.util.ArrayList;
import java.util.List;

class ObjectPool {
    private final List<MyObject> pool;
    private final int maxSize;

    public ObjectPool(int size) {
        this.maxSize = size;
        pool = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            pool.add(new MyObject(i));
        }
    }

    public MyObject acquire() {
        for (MyObject obj : pool) {
            if (!obj.isInUse()) {
                obj.setInUse(true);
                return obj;
            }
        }
        return null; // No available object
    }

    public void release(MyObject obj) {
        obj.setInUse(false);
    }
}

class MyObject {
    private final int id;
    private boolean inUse;

    public MyObject(int id) {
        this.id = id;
        this.inUse = false;
    }

    public int getId() {
        return id;
    }

    public boolean isInUse() {
        return inUse;
    }

    public void setInUse(boolean inUse) {
        this.inUse = inUse;
    }
}

public class Main {
    public static void main(String[] args) {
        ObjectPool pool = new ObjectPool(5);

        MyObject obj1 = pool.acquire();
        System.out.println("Acquired object with ID: " + obj1.getId());

        pool.release(obj1);
        System.out.println("Released object with ID: " + obj1.getId());

        MyObject obj2 = pool.acquire();
        System.out.println("Acquired object with ID: " + obj2.getId());
    }
}

File: pool/rust/object_pool.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug)]
struct Object {
    id: usize,
}

struct ObjectPool {
    pool: Vec<Rc<RefCell<Object>>>,
}

impl ObjectPool {
    fn new(size: usize) -> Self {
        let mut pool = Vec::with_capacity(size);
        for i in 0..size {
            pool.push(Rc::new(RefCell::new(Object { id: i })));
        }
        ObjectPool { pool }
    }

    fn acquire(&self) -> Option<Rc<RefCell<Object>>> {
        for obj in &self.pool {
            let mut obj_borrowed = obj.borrow_mut();
            if obj_borrowed.id != usize::MAX {
                let id = obj_borrowed.id;
                obj_borrowed.id = usize::MAX; // Marcar como em uso
                return Some(obj.clone());
            }
        }
        None
    }

    fn release(&self, obj: Rc<RefCell<Object>>) {
        let mut obj_borrowed = obj.borrow_mut();
        obj_borrowed.id = self.pool.iter().position(|x| Rc::ptr_eq(x, &obj)).unwrap();
    }
}

fn main() {
    let pool = ObjectPool::new(5);

    let obj1 = pool.acquire().expect("No object available");
    println!("Acquired object with ID: {}", obj1.borrow().id);

    pool.release(obj1.clone());
    println!("Released object with ID: {}", obj1.borrow().id);

    let obj2 = pool.acquire().expect("No object available");
    println!("Acquired object with ID: {}", obj2.borrow().id);
}
